File: 1342. Number of Steps to Reduce a Number to Zero/steps_zero.js
/**
    * Solved By: Gabriel E. Rodriguez Garcia
    * Date Completed: March 14, 2022
    * Date Optimized: July 11, 2023
    * Exersise #1342
    * Exercise Name: Number of steps to reduce a number to zero
    * Difficulty: Easy
    * Time Complexity: O(logn)
    * Space Complexity: O(n)
    * Language Used: JavaScript
    * Tags: Math, Bit Manipulation
*/

/**
 * Reduce a given number to 0 and counts the amount of steps taken
 * @param {number} num  A number to be reduced to 0
 * @return {number} Returns the amount of steps taken to reach 0
 */
var numberOfSteps = function(num) {
    //Initializes variable that will count steps.
    var steps = 0;
    //While nums is not zero keep reducing it
    while(num!=0){
        //If num is an even number divide num by 2
        if(num%2 == 0)
            num/=2;
        //If num is an odd number substract by 1
        else
            num-=1;
        //Increments steps by 1 each turn 
        steps++;
    }
    //Returns the amount of steps taken to reach 0
    return steps;
};

//Sample number
var num = 250;

//Prints to the console the amount of steps taken to reach 0
console.log(numberOfSteps(num));

File: 1342. Number of Steps to Reduce a Number to Zero/steps_zero.py
"""
    * Solved By: Gabriel E. Rodriguez Garcia
    * Date Completed: March 14, 2022
    * Date Optimized: July 11, 2023
    * Exersise #1342
    * Exercise Name: Number of steps to reduce a number to zero
    * Difficulty: Easy
    * Time Complexity: O(logn)
    * Space Complexity: O(n)
    * Language Used: Python
    * Tags: Math, Bit Manipulation
"""

"""
* Reduce a given number to 0 and counts the amount of steps taken
* Has a Parameter of nums, a number to be reduced to 0
* Returns the amount of steps taken to reach 0
"""
def numberOfSteps(num: int) -> int:
    # Initializes Variable that count steps
    steps = 0
    # While num is not equal to 0 keep looping
    while(num != 0):
        #if num is even then divide by 2
        if(num % 2 == 0):
            num /= 2
        else:
        #Else reduce the number by 1
            num-=1

        #Increment steps by 1
        steps+=1
    #Returns the amount of steps taken      
    return steps


class merge_strings:
    #Sample number
    num = 250
    #Prints the the console the amount of steps taken to reach 0
    print( str(numberOfSteps(num)) )
